File: Hamming_Code/Python/AlexMathew/HammingCode.py
"""
AUTHOR : Alex Mathew
EMAIL : alexmathew003@gmail.com 

There is a lot I can do to improve this implementation.

"""

"""For a given binary word, print out its corresponding Hamming Coded word"""

import math

def FindKeyBits(n):
  keyBits=[]
  pad = int(math.ceil(math.log(n)/math.log(2)))
  for i in xrange(1, n+1):
    b = "{0:b}".format(i)
    b=b.zfill(pad)
    keyBits.append(b)
  return keyBits

def IsPowerOf2(x):
  return ((x&(x-1))==0)

def XOR(a, b):
  return str(int(a)^int(b))

def FindParityCount(m):
  r = math.ceil((math.log(m)/math.log(2))+1)
  return int(r)

def CalculateParity(keyBits, data, parityTrack):
  parity = '0'
  pbCount = 0
  for elem in keyBits:
    if not IsPowerOf2(int(elem,2)):
      if elem[-parityTrack]=='1':
        parity = XOR(parity, data[pbCount-int(elem,2)])
    else:
      pbCount+=1
  return parity

def ComputeHamming(data):
  dataLen = len(data)
  parityCount = FindParityCount(dataLen)
  hammingLen = dataLen + parityCount
  keyBits = FindKeyBits(hammingLen)
  hammingCode = []
  dataTrack = -1
  parityTrack = 1
  for i in xrange(1, hammingLen+1):
    if IsPowerOf2(i):
      parity = CalculateParity(keyBits, data, parityTrack)
      hammingCode.append(parity)
      parityTrack+=1
    else:
      hammingCode.append(data[dataTrack])
      dataTrack-=1
  hammingCode.reverse()
  for i in hammingCode:
    print i, 


def main():
  print 'The inputs are passed into the HammingCode() function as a string \n'
  print '\nSample runs - '
  print '\n\nINPUT : 1101'
  print 'OUTPUT : '
  ComputeHamming('1101')
  print '\n\nINPUT : 11011101'
  print 'OUTPUT : '
  ComputeHamming('11011101')
  print '\n\n'
  data = raw_input("Enter the word to be encoded : ")
  print '\nThe Hamming encoded word for', data, 'is : ',
  ComputeHamming(data)
  
if __name__ == '__main__':
  main()

File: Hamming_Code/c++/lakshmanaram/HammingCode.cpp
#include<bits/stdc++.h>
using namespace std;
char XOR(char a,char b){
    return (char)((int)'0'+(((int)a-(int)'0')^((int)b-(int)'0')));
}
string ComputeHamming(string a){
    reverse(a.begin(),a.end());
    string ans = "";
    // ai represents the index of string, pow2 -> current power of 2
    for(int i=1, ai=0, pow2=1; ai<a.size(); i++){
        if(pow2 == i){
            pow2 *= 2;
            ans += "0";
        }else{
            ans += a[ai++];
            if(ans[i-1] == '0')
                continue;
            int temp_pow2 = 2;
            while(temp_pow2/2 <= i){
                if(i%temp_pow2 >= temp_pow2/2){
                    ans[(temp_pow2/2)-1] = XOR(ans[(temp_pow2/2)-1],ans[i-1]);
                }
                temp_pow2 *= 2;
            }
        }
    }
    reverse(ans.begin(),ans.end());
    return ans;
}
int main(){
    cout<<"The inputs are passed into the HammingCode() function as a string"<<endl;
    cout<<"\nSample runs - "<<endl;
    cout<<"\n\nINPUT : 1101"<<endl;
    cout<<"OUTPUT : "<<ComputeHamming("1101")<<endl;
    cout<<"\n\nINPUT : 11011101\n";
    cout<<"OUTPUT : "<<ComputeHamming("11011101")<<endl;
    cout<<"\n\nEnter the word to be encoded : ";
    string data;
    cin>>data;
    cout<<"\nThe Hamming encoded word for "<<data<<" is : "<<ComputeHamming(data);
}
